### DISCUSSION TITLE
How to implement a filter (annotation) applied on a sub-resource implementing an interface?

### DISCUSSION THREAD

**Discussion by fenix-hub** [quarkusio/quarkus#26020] [Q&A]
Hi everyone, I'm trying implementing a custom Filter but it is not working as expected.
I've got a Parent resource defined as an interface like so 
```java
@Path("/mainpath")
@RegisterRestClient
public interface IProvider extends Closeable {

	@POST
	@Produces(MediaType.APPLICATION_JSON)
	@Consumes(MediaType.APPLICATION_JSON)
	@Path("")
	public URI doSomething(
		...
	);
```
And a sub-resource implementing my interface
```java
public class Provider {

        @MyAnnotation
	@Override
	public URI doSomething(
		...
	);
```

I've created `MyAnnotation` like
```java
@NameBinding
@Documented
@Target({ElementType.METHOD, ElementType.TYPE})
@Retention(RetentionPolicy.RUNTIME)
public @interface MyAnnotation {
    int DEFAULT_CODE = 200;
    int DEFAULT_FILTER = 200;
    
    int code() default DEFAULT_CODE;
    int filter() default DEFAULT_FILTER;
}
```
and a ContainerResponseFilter to handle this with name binding
```java
@MyAnnotation 
@Provider
@Priority(Priorities.HEADER_DECORATOR)
public class MyAnnotationFilter implements ContainerResponseFilter {

    @Context
    private ResourceInfo info;

    @Override
    public void filter(ContainerRequestContext req, ContainerResponseContext res) throws IOException {
        MyAnnotation responseStatus = getAnnotation(info.getResourceMethod());
        etc ...
    }

    private MyAnnotation getAnnotation(Method resourceMethod) {
        String methodName = resourceMethod.getName();
        Class<?>[] paramTypes = resourceMethod.getParameterTypes();
        Class<?> declaringClass = resourceMethod.getDeclaringClass();
        Method declaredMethod;
        try {
            declaredMethod = declaringClass.getDeclaredMethod(methodName, paramTypes);
        } catch (NoSuchMethodException e) {
            return null;
        }
        return declaredMethod.getAnnotation(MyAnnotation.class);
    }
}
```

If `@MyAnnotation` is applyed to the method in the class implementing the interface (like the example I'm showing) it is completely ignored. But if the annotation is applied inside the interface itself, it works.
Since other annotations, like `@RolesAllowed` or `@DenyAll` are working in the sub-resource, i'm expecting my annotation to work too. Is there something I'm missing?

**Comment by fenix-hub** [comment]
I also tried with an Interceptor instead of a Filter, it doesn't work.
Looks like everything which is *not* annotated with `@Path` is ignored. But why other annotations work?

**Comment by NMichas** [comment]
Just to add, I tested this on 2.11.1.Final and the behaviour is still the same. I, similarly, try to implement a dynamic filter and having the custom annotation on the Interface isn't an option.

If I duplicate the `@Path` annotation on the interface as well as in the implementation class, it's ugly, but it works.

@fenix-hub did you manage to find a proper solution here?

**Reply by fenix-hub** [reply]
@NMichas hi!
Sadly the only solution I've found is to develop a Quarkus extension just for the annotation definition.
The extension is needed in order to register the Interceptor through its annotation directly in the Quarkus runtime, so it is executed by the framework itself when implemented.
I started digging a little bit through the quarkus issues and I've found this some months ago:
https://github.com/quarkusio/quarkus/issues/2280
The context is different but the issue is completely relatable to our case.
Then I've found this answer:
https://github.com/quarkusio/quarkus/issues/2280#issuecomment-487703164
Which explains WHY the filter is ignored, and it makes sense since everything which is not annotated with `@Path` is removed.
Sadly the op's solution was still not useful for me, since the issue was not the interface itself, but applying a filter on the class implementing it.. so I dug up some more. The question which bothered me was always "Why doesn't *my* annotation work in this case, but security annotations like *`@Authenticated`* and *`@DenyAll`* worked in the same exact context?" There was no way I was doing something wrong, so the problem was related to something *inside* the framework... 
and found this in the source code: https://github.com/quarkusio/quarkus/blob/40db70f0f160c55b4b426e11775dc4e6f80ee0a1/extensions/security/deployment/src/main/java/io/quarkus/security/deployment/SecurityProcessor.java#L389
Security annotations are explicitely added in a `@BuildStep`, so the framework was able to resolve them.
I thought that creating an extesion which adds a `@BuildStep` that configures my `@MyAnnotation` could work, but didn't experiment it in the end, so the solution is pretty much theoretical... but should work.

**Reply by fenix-hub** [reply]
Here's the actual starting point of what should be implemented in order to test this solution.
https://quarkus.io/guides/cdi-integration#additional_bean_build_item
The reason why I didn't try is just that I was losing a lot of time digging up the solution so I just dropped the annotation itself and moved on.

**Reply by NMichas** [reply]
Thanks a lot for your time preparing such a detailed reply, it makes sense now.

I had a look on developing custom extensions, definitely an overkill for me right now. For the time being I'll just stay with duplicate annotations :( unless devs have a better proposal :)

**Reply by fenix-hub** [reply]
You are welcome

> I had a look on developing custom extensions, definitely an overkill for me right now. For the time being I'll just stay with duplicate annotations :( unless devs have a better proposal :)

https://github.com/quarkusio/quarkus/issues/26075#issuecomment-1167066051
